src/chainrules.jl

```diff
@@ -2,6 +2,48 @@ ChainRulesCore.@scalar_rule(xlogx(x::Real), (1 + log(x),))
 ChainRulesCore.@scalar_rule(xlogy(x::Real, y::Real), (log(y), x / y,))
 ChainRulesCore.@scalar_rule(xlog1py(x::Real, y::Real), (log1p(y), x / (1 + y),))
 
+function ChainRulesCore.frule((_, Δx), ::typeof(xexpx), x::Real)
+    e = exp(x)
+    if iszero(e)
+        Ω = e
+        ΔΩ = e * Δx
+    else
+        Ω = x * e
+        ΔΩ = (1 + x) * e * Δx
+    end
+    return Ω, ΔΩ
+end
+
+function ChainRulesCore.rrule(::typeof(xexpx), x::Real)
+    e = exp(x)
+    Ω = iszero(e) ? e : x * e
+    xexpx_pullback(Δ) = (ChainRulesCore.NoTangent(), iszero(e) ? e * Δ : (1 + x) * e * Δ)
+    return Ω, xexpx_pullback
+end
+
+function ChainRulesCore.frule((_, Δx, Δy), ::typeof(xexpy), x::T, y::T) where {T<:Real}
+    e = exp(y)
+    if iszero(e) && !isnan(x)
+        Ω = e
+        ΔΩ = e * Δx + zero(Ω * Δy)
+    else
+        Ω = x * e
+        ΔΩ = e * Δx + Ω * Δy
+    end
+    return Ω, ΔΩ
+end
+
+function ChainRulesCore.rrule(::typeof(xexpy), x::T, y::T) where {T<:Real}
+    e = exp(y)
+    Ω = iszero(e) && !isnan(x) ? e : x * e
+    xexpy_pullback(Δ) = if iszero(e) && !isnan(x)
+        (ChainRulesCore.NoTangent(), e * Δ, zero(Ω * Δ))
+    else
+        (ChainRulesCore.NoTangent(), e * Δ, Ω * Δ)
+    end
```

Comment:
Again, I'd suggest to simplify and improve readability by using
```suggestion
    expy = exp(y)
    Ω = iszero(expy) && !isnan(x) ? expy : x * expy
    function xexpy_pullback(ΔΩ)
        Δx = expy * ΔΩ
        Δy = Ω * ΔΩ
        return ChainRulesCore.NoTangent(), Δx, Δy
    end
```